### PR TITLE
SSR-2366: Edit Table component proto

### PR DESCRIPTION
**Pull request by voduchau**

### CODE REVIEW

File: src/base/constant.ts [hunk 1]
```diff
@@ -10,4 +10,12 @@ export const ERROR_MESSAGE = {
     IS_NOT_ARRAY: "'selectedIndex' property is not array",
     IS_NOT_NUMBER: "'selectedIndex' property is not number",
   },
+  COLUMNS: {
+    IS_NOT_ARRAY: "'columns' property is not array.",
+    FIELD_REQUIRED: "'field' property is not specified in columns.",
+    FIELD_UNIQUE: "'field' property is not unique in columns.",
+  },
+  DATA_TABLE: {
+    IS_NOT_ARRAY: "data' property is not array.",
```
Comment (TomokoMiyake) [inline src/base/constant.ts:19]:
```suggestion
    IS_NOT_ARRAY: "'data' property is not array.",
```

**Review by TomokoMiyake — CHANGES_REQUESTED** [review]

File: src/table/index.stories.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  },
+  parameters: {
+    actions: {
+      handles: ["change"],
+    },
+  },
+};
+
+const renderAge = (dataCell, dataRow) => {
+  const text = new TextArea({ value: dataCell });
```
Comment (TomokoMiyake) [inline src/table/index.stories.js:28]:
```suggestion
const renderAge = (cellData, rowData) => {
  const text = new TextArea({ value: cellData });
```

File: src/table/type.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+export declare type Column = {
+  title?: string;
+  field: string;
+  requiredIcon?: boolean;
+  visible?: boolean;
+  render?: Render;
+};
+export declare type Render = (
+  dataCell: any,
+  dataRow: object,
```
Comment (TomokoMiyake) [inline src/table/type.ts:12]:
```suggestion
  cellData: any,
  rowData: object,
```

**Review by TomokoMiyake — CHANGES_REQUESTED** [review]

File: src/table/index.ts [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+  dispatchCustomEvent,
+  CustomEventDetail,
+  createStyleOnHeader,
+} from "../base/kuc-base";
+import { visiblePropConverter } from "../base/converter";
+import {
+  validateProps,
+  throwErrorAfterUpdateComplete,
+  validateArrayProperty,
+  validateFieldRequiedInColumnTable,
```
Comment (TomokoMiyake) [inline src/table/index.ts:15]:
```suggestion
  validateFieldRequiredInColumnTable,
```

File: src/table/index.ts [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+        return false;
+      }
+      return true;
+    }
+
+    private _getErrorMessageWhenValidateColumns() {
+      if (!validateArrayProperty(this.columns)) {
+        return ERROR_MESSAGE.COLUMNS.IS_NOT_ARRAY;
+      }
+      if (!validateFieldRequiedInColumnTable(this.columns)) {
```
Comment (TomokoMiyake) [inline src/table/index.ts:90]:
```suggestion
      if (!validateFieldRequiredInColumnTable(this.columns)) {
```

**Review by TomokoMiyake — CHANGES_REQUESTED** [review]

**Review by TomokoMiyake — APPROVED** [review]
LGTM!

File: src/base/validator.ts [hunk 6]
```diff
@@ -121,3 +121,30 @@ export async function throwErrorAfterUpdateComplete(
   await _this.updateComplete;
   throw new Error(message);
 }
+
+export const validateFieldRequiredInColumnTable = (columns: object[]) => {
+  for (let i = 0; i < columns.length; i++) {
+    const column = columns[i];
+    const containField = "field" in column;
```
Comment (nmanhit) [inline src/base/validator.ts:128]:
```
if (!column.hasOwnProperty("field")) return false;
```
The ``hasOwnProperty`` will be faster than the ``in``

File: src/base/validator.ts [hunk 7]
```diff
@@ -121,3 +121,30 @@ export async function throwErrorAfterUpdateComplete(
...
+    if (!containField) return false;
+  }
+
+  return true;
+};
+
+export const validateFieldUniqueInColumnTable = (columns: object[]) => {
+  const valueArr = columns.map(function (item: any) {
```
Comment (nmanhit) [inline src/base/validator.ts:136]: the code written in the short format
```const valueArr = columns.map((item) => item.field);```

File: src/base/validator.ts [hunk 8]
```diff
@@ -121,3 +121,30 @@ export async function throwErrorAfterUpdateComplete(
...
+  });
+  const isDuplicate = valueArr.some(function (item, idx) {
+    return valueArr.indexOf(item) !== idx;
+  });
+
+  return isDuplicate;
+};
+
+export const validateArrayProperty = (data: object[]) => {
+  if (!Array.isArray(data)) return false;
```
Comment (nmanhit) [inline src/base/validator.ts:147]: ```return Array.isArray(data)```

File: src/table/index.ts [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+            ${this.data.map((dataRow: object, rowIndex: number) => {
+              return this._getTableRowTemplate(dataRow, rowIndex);
+            })}
+          </tbody>
+        </table>
+      `;
+    }
+
+    private _getTableHeaderTemplate() {
+      return html`
```
Comment (nmanhit) [inline src/table/index.ts:118]: return this.columns.length === 0 ? null : html` .....`

File: src/table/style.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        color: #ffffff;
+        height: 40px;
+        padding: 4px 8px;
+        text-align: left;
+        white-space: nowrap;
+    }
+    .kuc-table__table__header__cell[hidden] {
+        display: none;
+    }
+    .kuc-table__table__header__cell__label {
```
Comment (nmanhit) [inline src/table/style.ts:58]: We don't use this class

File: src/table/index.ts [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+        return ERROR_MESSAGE.COLUMNS.FIELD_REQUIRED;
+      }
+      if (validateFieldUniqueInColumnTable(this.columns)) {
+        return ERROR_MESSAGE.COLUMNS.FIELD_UNIQUE;
+      }
+      return "";
+    }
+
+    render() {
+      return html`
```
Comment (nmanhit) [inline src/table/index.ts:106]: return !this.columns || this.columns.length < 1
        ? html`
            <table class="kuc-table__table">
              <caption
                class="kuc-table__table__label"
                ?hidden="${!this.label}"
              >
                ${this.label}
              </caption>
            </table>
          `
        : html` .....`

File: src/table/index.stories.js [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    ],
+    value: cellData,
+  });
+  return radioButton;
+};
+
+const renderMultiChoice = (cellData) => {
+  const multichoice = new MultiChoice({
+    items: [
+      { label: "sample1", value: "sample1" },
```
Comment (nmanhit) [inline src/table/index.stories.js:92]:
```
{ label: "Orange", value: "orange" },
{ label: "Banana", value: "banana" },
{ label: "Tomato", value: "tomato" },
```

File: src/table/index.ts [hunk 13]
```diff
@@ -375,6 +375,18 @@ let exportTable;
       return buttonAction;
     }
 
+    private _getErrorValidateColumnsAndData() {
+      let errorMessage = "";
+      const errorColumns = this._getErrorMessageWhenValidateColumns();
+      if (errorColumns) {
+        errorMessage = errorColumns;
```
Comment (nmanhit) [inline src/table/index.ts:382]: return errorColumns

File: src/table/index.ts [hunk 14]
```diff
@@ -375,6 +375,18 @@ let exportTable;
...
+      }
+      if (!validateArrayProperty(this.data)) {
+        errorMessage = ERROR_MESSAGE.DATA_TABLE.IS_NOT_ARRAY;
```
Comment (nmanhit) [inline src/table/index.ts:385]: return ERROR_MESSAGE.DATA_TABLE.IS_NOT_ARRAY;

File: src/table/index.ts [hunk 15]
```diff
@@ -375,6 +375,18 @@ let exportTable;
...
+      }
+      return errorMessage;
```
Comment (nmanhit) [inline src/table/index.ts:387]: return "";

**Pull request merged**